
# NBA 3-Point Statistics

## Introduction

There has never been a single metric to determine the best NBA 3-Point shooter. Most fans consider two numbers, 3-Point Percentage, and 3 Pointers Made. Is there a way to combine these two categories into one?

This Jupyter Notebook reveals a new metric that will rank NBA 3-Point shooters using one number. I considered different ways of combining 3-Point Percentage and 3-Pointers Made before coming up with an empirical solution.


#### References

https://www.kaggle.com/drgilermo/nba-players-stats <br>
https://www.basketball-reference.com/leagues/NBA_2018_totals.html

#### Copyright

Corey J Wade<br>
May 28,2018

This Jupyter Notebook and the statistics within may be redistributed provided that Corey J Wade is given credit.


## Import NBA Statistics

The following csv file is taken from https://www.kaggle.com/drgilermo/nba-players-stats. When I downloaded the file, it contained standard statistics through 2017. Dr. Guillermo scraped it from https://www.basketball-reference.com/. 

#### NBA Stats Through 2017

In [1]:
# import pandas
import pandas as pd

# open file as dataframe
df_2017 = pd.read_csv('Seasons_Stats.csv')

# display first five rows
df_2017.head()

Unnamed: 0    Year           Player  Pos   Age   Tm     G  GS  MP  PER  \
0           0  1950.0  Curly Armstrong  G-F  31.0  FTW  63.0 NaN NaN  NaN   
1           1  1950.0     Cliff Barker   SG  29.0  INO  49.0 NaN NaN  NaN   
2           2  1950.0    Leo Barnhorst   SF  25.0  CHS  67.0 NaN NaN  NaN   
3           3  1950.0       Ed Bartels    F  24.0  TOT  15.0 NaN NaN  NaN   
4           4  1950.0       Ed Bartels    F  24.0  DNN  13.0 NaN NaN  NaN   

   ...      FT%  ORB  DRB  TRB    AST  STL  BLK  TOV     PF    PTS  
0  ...    0.705  NaN  NaN  NaN  176.0  NaN  NaN  NaN  217.0  458.0  
1  ...    0.708  NaN  NaN  NaN  109.0  NaN  NaN  NaN   99.0  279.0  
2  ...    0.698  NaN  NaN  NaN  140.0  NaN  NaN  NaN  192.0  438.0  
3  ...    0.559  NaN  NaN  NaN   20.0  NaN  NaN  NaN   29.0   63.0  
4  ...    0.548  NaN  NaN  NaN   20.0  NaN  NaN  NaN   27.0   59.0  

[5 rows x 53 columns]

Statistics were not widely computed before the modern era, hence the null values.

In [2]:
# delete unnecessary column
del df_2017['Unnamed: 0']

# display last five rows
df_2017.tail()

Year             Player Pos   Age   Tm     G    GS      MP   PER  \
24686  2017.0        Cody Zeller  PF  24.0  CHO  62.0  58.0  1725.0  16.7   
24687  2017.0       Tyler Zeller   C  27.0  BOS  51.0   5.0   525.0  13.0   
24688  2017.0  Stephen Zimmerman   C  20.0  ORL  19.0   0.0   108.0   7.3   
24689  2017.0        Paul Zipser  SF  22.0  CHI  44.0  18.0   843.0   6.9   
24690  2017.0        Ivica Zubac   C  19.0  LAL  38.0  11.0   609.0  17.0   

         TS%  ...      FT%    ORB    DRB    TRB   AST   STL   BLK   TOV  \
24686  0.604  ...    0.679  135.0  270.0  405.0  99.0  62.0  58.0  65.0   
24687  0.508  ...    0.564   43.0   81.0  124.0  42.0   7.0  21.0  20.0   
24688  0.346  ...    0.600   11.0   24.0   35.0   4.0   2.0   5.0   3.0   
24689  0.503  ...    0.775   15.0  110.0  125.0  36.0  15.0  16.0  40.0   
24690  0.547  ...    0.653   41.0  118.0  159.0  30.0  14.0  33.0  30.0   

          PF    PTS  
24686  189.0  639.0  
24687   61.0  178.0  
24688   17.0   23.0  
24689   78.0  240.0  
24690   66.0  284.0  

[5 rows x 52 columns]

As expected, all visible columns are full of data.

#### 2018 NBA Stats

The 2018 NBA season recently finished. I used the same link, https://www.basketball-reference.com/, to scrape the 2018 statistics.

In [3]:
# read html file
df_2018, = pd.read_html("https://www.basketball-reference.com/leagues/NBA_2018_totals.html", header=0)

# convert to csv file
df_2018.to_csv("tp_2016.csv", index=False)

# display first five rows
df_2018.head()

Rk         Player Pos Age   Tm   G  GS    MP   FG  FGA  ...    FT%  ORB  \
0  1   Alex Abrines  SG  24  OKC  75   8  1134  115  291  ...   .848   26   
1  2     Quincy Acy  PF  27  BRK  70   8  1359  130  365  ...   .817   40   
2  3   Steven Adams   C  24  OKC  76  76  2487  448  712  ...   .557  384   
3  4    Bam Adebayo   C  20  MIA  69  19  1368  174  340  ...   .721  118   
4  5  Arron Afflalo  SG  32  ORL  53   3   682   65  162  ...   .846    4   

   DRB  TRB  AST STL BLK  TOV   PF   PTS  
0   88  114   28  38   8   25  124   353  
1  216  256   57  33  29   60  149   411  
2  301  685   88  92  78  128  215  1056  
3  263  381  101  32  41   66  138   477  
4   62   66   30   4   9   21   56   179  

[5 rows x 30 columns]

Since there is no column for year, I will add one. Since I plan to concatenate the two dataframes, their columns need to match.

In [4]:
# delete unnecessary column
del df_2018['Rk']

# add column for year, place at index 0
df_2018.insert(0, 'Year', 2018.0)

# display last five rows
df_2018.tail()

Year        Player Pos Age   Tm   G  GS   MP   FG  FGA ...    FT% ORB  \
685  2018.0  Tyler Zeller   C  28  BRK  42  33  703  125  229 ...   .667  63   
686  2018.0  Tyler Zeller   C  28  MIL  24   1  406   62  105 ...   .895  47   
687  2018.0   Paul Zipser  SF  23  CHI  54  12  824   81  234 ...   .731  13   
688  2018.0    Ante Zizic   C  21  CLE  32   2  214   49   67 ...   .724  24   
689  2018.0   Ivica Zubac   C  20  LAL  43   0  410   61  122 ...   .765  45   

     DRB  TRB AST STL BLK TOV  PF  PTS  
685  131  194  28   8  21  35  78  300  
686   64  111  19   7  14  12  48  141  
687  118  131  46  20  15  43  86  218  
688   36   60   5   2  13  11  30  119  
689   78  123  25   8  15  26  47  161  

[5 rows x 30 columns]

#### Concatenating Dataframes

Since the dataframes have a different number of columns, and I clearly do not need all of the columns since my focus is on 3-pointers, I will select the relevant columns before concatenating.

In [5]:
# select relevant columns
tp_2017 = df_2017[['Year', 'Tm', 'Player', 'G','MP', 'PTS', '3P', '3PA', '3P%']]
tp_2018 = df_2018[['Year', 'Tm', 'Player', 'G','MP', 'PTS', '3P', '3PA', '3P%']]

# concatenate dataframes
tp = pd.concat([tp_2017, tp_2018], ignore_index=True, )

# show last five rows
tp.tail()

Year   Tm        Player   G   MP  PTS  3P  3PA   3P%
25376  2018.0  BRK  Tyler Zeller  42  703  300  10   26  .385
25377  2018.0  MIL  Tyler Zeller  24  406  141   0    2  .000
25378  2018.0  CHI   Paul Zipser  54  824  218  37  110  .336
25379  2018.0  CLE    Ante Zizic  32  214  119   0    0   NaN
25380  2018.0  LAL   Ivica Zubac  43  410  161   0    1  .000

## Editing Dataframe

#### Column Consistency

In [6]:
# display column info
tp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25381 entries, 0 to 25380
Data columns (total 9 columns):
Year      25314 non-null float64
Tm        25314 non-null object
Player    25314 non-null object
G         25314 non-null object
MP        24828 non-null object
PTS       25314 non-null object
3P        19617 non-null object
3PA       19617 non-null object
3P%       16041 non-null object
dtypes: float64(1), object(8)
memory usage: 1.7+ MB


With the exception of 'Year', the data has not been rendered as numbers. They must be converted to floats for mathematical operations.

In [7]:
# convert numeric columns to decimals
tp.G = pd.to_numeric(tp.G, errors='coerce')
tp.MP = pd.to_numeric(tp.MP, errors='coerce')
tp.PTS = pd.to_numeric(tp.PTS, errors='coerce')
tp['3P'] = pd.to_numeric(tp['3P'], errors='coerce')
tp['3PA'] = pd.to_numeric(tp['3PA'], errors='coerce')
tp['3P%'] = pd.to_numeric(tp['3P%'], errors='coerce')

# check columns
tp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25381 entries, 0 to 25380
Data columns (total 9 columns):
Year      25314 non-null float64
Tm        25314 non-null object
Player    25314 non-null object
G         25288 non-null float64
MP        24802 non-null float64
PTS       25288 non-null float64
3P        19591 non-null float64
3PA       19591 non-null float64
3P%       16015 non-null float64
dtypes: float64(7), object(2)
memory usage: 1.7+ MB


#### Minimum Requirements

It's not necessary to examine data from all players. For instance, if a player was never recorded as taking a 3-pointer, he should be excluded from the dataframe. The same holds for a player who only took one 3-pointer and made it. The purpose of the minimum requirements is to eliminate non-3-point shooters and low outliers.

In [8]:
# choose players with more than 20 3's
tp = tp[(tp['3P'] > 10)]

# chosee players with more than 320 mintes played
tp = tp[(tp['MP'] > 320)]

# choose players with at least 41 games
tp = tp[(tp['G'] > 41)]

# display last 5 rows
tp.tail()

Year   Tm          Player     G      MP    PTS     3P    3PA    3P%
25368  2018.0  TOR    Delon Wright  69.0  1433.0  555.0   56.0  153.0  0.366
25371  2018.0  IND       Joe Young  53.0   558.0  207.0   25.0   66.0  0.379
25372  2018.0  GSW      Nick Young  80.0  1393.0  581.0  123.0  326.0  0.377
25373  2018.0  IND  Thaddeus Young  81.0  2607.0  955.0   58.0  181.0  0.320
25378  2018.0  CHI     Paul Zipser  54.0   824.0  218.0   37.0  110.0  0.336

In [9]:
tp.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5928 entries, 5734 to 25378
Data columns (total 9 columns):
Year      5928 non-null float64
Tm        5928 non-null object
Player    5928 non-null object
G         5928 non-null float64
MP        5928 non-null float64
PTS       5928 non-null float64
3P        5928 non-null float64
3PA       5928 non-null float64
3P%       5928 non-null float64
dtypes: float64(7), object(2)
memory usage: 463.1+ KB


## New 3-point Statistics

### Expected Minutes

This first group of statistics computes the number of minutes players are on the court before attemping and making 3's.

#### 1) AM3A : Average Minutes per 3-point Attempt

A player's Average Minutes per 3-Point Attempt is total minutes played divided by total 3-pointers attempted.

In [10]:
# define new column, AM3A: Average Minutes per 3-point Attempt 
tp['AM3A'] = tp['MP'] / tp['3PA']

# show last five entrants
tp.tail()

Year   Tm          Player     G      MP    PTS     3P    3PA    3P%  \
25368  2018.0  TOR    Delon Wright  69.0  1433.0  555.0   56.0  153.0  0.366   
25371  2018.0  IND       Joe Young  53.0   558.0  207.0   25.0   66.0  0.379   
25372  2018.0  GSW      Nick Young  80.0  1393.0  581.0  123.0  326.0  0.377   
25373  2018.0  IND  Thaddeus Young  81.0  2607.0  955.0   58.0  181.0  0.320   
25378  2018.0  CHI     Paul Zipser  54.0   824.0  218.0   37.0  110.0  0.336   

            AM3A  
25368   9.366013  
25371   8.454545  
25372   4.273006  
25373  14.403315  
25378   7.490909

#### 2) EM3A : Expected Minutes before 3-point Attempt

The expected value of a continuous interval of time is typically at the halfway mark. Will Nick Young (listed above) take a 3 once he checks in, or after 4.27 minutes? His most likely value is halfway between, at 2.135 minutes. This is his expected minutes played before attempting a 3. 

In [11]:
# define new column, EM3A: Expected Minutes before 3-point Attempt
tp['EM3A'] = tp['AM3A'] / 2

# sort dataframe by new category
tp_EM3A = tp.sort_values('EM3A', ascending=True)

# view players who attempt 3s faster than anyone in NBA history
tp_EM3A.head(20)

Year   Tm              Player     G      MP     PTS     3P    3PA  \
25279  2018.0  ORL   Marreese Speights  52.0   675.0   402.0   86.0  233.0   
25119  2018.0  TOR          C.J. Miles  70.0  1337.0   699.0  164.0  454.0   
23633  2016.0  GSW       Stephen Curry  79.0  2700.0  2375.0  402.0  886.0   
23463  2015.0  DAL  Charlie Villanueva  64.0   678.0   403.0   83.0  221.0   
24842  2018.0  GSW       Stephen Curry  51.0  1631.0  1346.0  212.0  501.0   
24217  2017.0  MEM        Troy Daniels  67.0  1183.0   551.0  138.0  355.0   
24216  2017.0  GSW       Stephen Curry  79.0  2638.0  1999.0  324.0  789.0   
23002  2015.0  TOT        Troy Daniels  47.0   397.0   176.0   43.0  118.0   
24290  2017.0  HOU         Eric Gordon  75.0  2323.0  1217.0  246.0  661.0   
25130  2018.0  CHO          Malik Monk  63.0   854.0   421.0   83.0  243.0   
24874  2018.0  MIA     Wayne Ellington  77.0  2041.0   864.0  227.0  579.0   
24932  2018.0  HOU        James Harden  72.0  2551.0  2191.0  265.0  722.0   
24844  2018.0  PHO        Troy Daniels  79.0  1622.0   703.0  183.0  458.0   
24915  2018.0  HOU         Eric Gordon  69.0  2154.0  1243.0  218.0  608.0   
21504  2012.0  NYK         Steve Novak  54.0  1020.0   477.0  133.0  282.0   
19222  2008.0  DEN          J.R. Smith  74.0  1421.0   907.0  157.0  390.0   
24024  2016.0  PHO     Mirza Teletovic  79.0  1686.0   965.0  181.0  460.0   
6575   1982.0  GSW         Joe Hassett  68.0   787.0   390.0   71.0  214.0   
24684  2017.0  LAL          Nick Young  60.0  1556.0   791.0  170.0  421.0   
24609  2017.0  MIL     Mirza Teletovic  70.0  1133.0   451.0  104.0  305.0   

         3P%      AM3A      EM3A  
25279  0.369  2.896996  1.448498  
25119  0.361  2.944934  1.472467  
23633  0.454  3.047404  1.523702  
23463  0.376  3.067873  1.533937  
24842  0.423  3.255489  1.627745  
24217  0.389  3.332394  1.666197  
24216  0.411  3.343473  1.671736  
23002  0.364  3.364407  1.682203  
24290  0.372  3.514372  1.757186  
25130  0.342  3.514403  1.757202  
24874  0.392  3.525043  1.762522  
24932  0.367  3.533241  1.766620  
24844  0.400  3.541485  1.770742  
24915  0.359  3.542763  1.771382  
21504  0.472  3.617021  1.808511  
19222  0.403  3.643590  1.821795  
24024  0.393  3.665217  1.832609  
6575   0.332  3.677570  1.838785  
24684  0.404  3.695962  1.847981  
24609  0.341  3.714754  1.857377

Statistical Notes<ul>
    <li> Many players on the list come off the bench. EM3A does not distinguish between starters and reserves.</li>
    <li> Most top performers are from the last few years, due to the meteoric rise of NBA 3-pointers. </li>
     <li> Joe Hassett from 1982 is a shocker. </li>
    </ul>

#### 3) AM3P : Average Minutes per 3-Pointer

AMP3 is the player's total minutes divided by the number of 3-pointers made. This statistic is more interesting because we are now looking at makes instead of just attempts.

In [12]:
# define new column, AM3P: Average Minutes per 3-pointer
tp['AM3P'] = tp['MP']/tp['3P']

# show last five rows
tp.tail()

Year   Tm          Player     G      MP    PTS     3P    3PA    3P%  \
25368  2018.0  TOR    Delon Wright  69.0  1433.0  555.0   56.0  153.0  0.366   
25371  2018.0  IND       Joe Young  53.0   558.0  207.0   25.0   66.0  0.379   
25372  2018.0  GSW      Nick Young  80.0  1393.0  581.0  123.0  326.0  0.377   
25373  2018.0  IND  Thaddeus Young  81.0  2607.0  955.0   58.0  181.0  0.320   
25378  2018.0  CHI     Paul Zipser  54.0   824.0  218.0   37.0  110.0  0.336   

            AM3A      EM3A       AM3P  
25368   9.366013  4.683007  25.589286  
25371   8.454545  4.227273  22.320000  
25372   4.273006  2.136503  11.325203  
25373  14.403315  7.201657  44.948276  
25378   7.490909  3.745455  22.270270

#### 4) EMB3 : Expected Minutes Before a 3

This is the best statistic of the group. It's how long a player is expected to be on the court before making a 3. As before, Average Minutes per 3-pointer is simply divided by two.

In [13]:
# define new column, EM3P: Expected Minutes before 3-pointer
tp['EM3P'] = tp['AM3P'] / 2

# sort dataframe by new category
tp_EM3P = tp.sort_values('EM3P', ascending=True)

# display top twenty seasons of all-time
tp_EM3P.head(20)

Year   Tm              Player     G      MP     PTS     3P    3PA  \
23633  2016.0  GSW       Stephen Curry  79.0  2700.0  2375.0  402.0  886.0   
21504  2012.0  NYK         Steve Novak  54.0  1020.0   477.0  133.0  282.0   
24842  2018.0  GSW       Stephen Curry  51.0  1631.0  1346.0  212.0  501.0   
25279  2018.0  ORL   Marreese Speights  52.0   675.0   402.0   86.0  233.0   
23634  2016.0  CHO        Troy Daniels  43.0   476.0   242.0   59.0  122.0   
24216  2017.0  GSW       Stephen Curry  79.0  2638.0  1999.0  324.0  789.0   
25119  2018.0  TOR          C.J. Miles  70.0  1337.0   699.0  164.0  454.0   
23463  2015.0  DAL  Charlie Villanueva  64.0   678.0   403.0   83.0  221.0   
24217  2017.0  MEM        Troy Daniels  67.0  1183.0   551.0  138.0  355.0   
24844  2018.0  PHO        Troy Daniels  79.0  1622.0   703.0  183.0  458.0   
24874  2018.0  MIA     Wayne Ellington  77.0  2041.0   864.0  227.0  579.0   
19222  2008.0  DEN          J.R. Smith  74.0  1421.0   907.0  157.0  390.0   
23000  2015.0  GSW       Stephen Curry  80.0  2613.0  1900.0  286.0  646.0   
24684  2017.0  LAL          Nick Young  60.0  1556.0   791.0  170.0  421.0   
20644  2011.0  OKC        Daequan Cook  43.0   598.0   239.0   65.0  154.0   
23002  2015.0  TOT        Troy Daniels  47.0   397.0   176.0   43.0  118.0   
24024  2016.0  PHO     Mirza Teletovic  79.0  1686.0   965.0  181.0  460.0   
24290  2017.0  HOU         Eric Gordon  75.0  2323.0  1217.0  246.0  661.0   
25051  2018.0  CLE         Kyle Korver  73.0  1574.0   672.0  164.0  376.0   
24932  2018.0  HOU        James Harden  72.0  2551.0  2191.0  265.0  722.0   

         3P%      AM3A      EM3A      AM3P      EM3P  
23633  0.454  3.047404  1.523702  6.716418  3.358209  
21504  0.472  3.617021  1.808511  7.669173  3.834586  
24842  0.423  3.255489  1.627745  7.693396  3.846698  
25279  0.369  2.896996  1.448498  7.848837  3.924419  
23634  0.484  3.901639  1.950820  8.067797  4.033898  
24216  0.411  3.343473  1.671736  8.141975  4.070988  
25119  0.361  2.944934  1.472467  8.152439  4.076220  
23463  0.376  3.067873  1.533937  8.168675  4.084337  
24217  0.389  3.332394  1.666197  8.572464  4.286232  
24844  0.400  3.541485  1.770742  8.863388  4.431694  
24874  0.392  3.525043  1.762522  8.991189  4.495595  
19222  0.403  3.643590  1.821795  9.050955  4.525478  
23000  0.443  4.044892  2.022446  9.136364  4.568182  
24684  0.404  3.695962  1.847981  9.152941  4.576471  
20644  0.422  3.883117  1.941558  9.200000  4.600000  
23002  0.364  3.364407  1.682203  9.232558  4.616279  
24024  0.393  3.665217  1.832609  9.314917  4.657459  
24290  0.372  3.514372  1.757186  9.443089  4.721545  
25051  0.436  4.186170  2.093085  9.597561  4.798780  
24932  0.367  3.533241  1.766620  9.626415  4.813208

EM3P Statistical Notes:<ul>
    <li> EMB3, like 3-Point Percentage, levels the playing field between starters and reserves.</li> 
    <li> More restrictive minimum requirements would eliminate many players from this list. </li>
     <li> Steph Curry's legendary 2016 MVP season is a clear # 1. </li>
    </ul>

### 3-Point Advantage

Although the 3-point statistics above are compelling, they do not a provide a single metric to rank 3-point shooters. This is where 3-point Advantage comes in. Simply put, 3-Point Advantage adds the net gain per 3-pointer made, and subtracts the net loss per 3-pointer missed. 

#### Points Per Possession

Computing net gain and net loss depends on the expected value. Should the expected value be points per possession? Or points per field goal attempt? There is no clear answer. I have decided to employ the classical strategy of assuming that NBA teams average 1.0 points per possession, very close to current averages. The expected value will be 1.0 points per possession.

#### Computation

The computation is determined by expected value. When a player makes a 3-pointer, the team gains more than the expected value. How much more? An additional 3 points minus the expected value. When a player misses a 3-pointer, the team does worse than the expected value. How much worse? They lose the expected value.

More concisely, to compute 3-Point Advantage, each 3-pointer made is multipled by 3 minus the expected value. Each 3-pointer missed is multiped by the expected value. 


In [14]:
# 3PAd, 3-Point Advantage

# Compute 3PG, 3-pointers per Game
tp['3PG']=tp['3P']/tp['G']

# Compute 3PAG, 3-point Attempts per Game
tp['3PAG']= tp['3PA']/tp['G']

# Compute 3PMi, 3-point Misses per Game
tp['3PMiG']=tp['3PAG']-tp['3PG']

# Declare expected value
ev = 1.0
                          
# Compute 3PAd, 3-point Advantage
tp['3PAd']=tp['3PG'] * (3 - ev) - tp['3PMiG'] * ev

# (3 - ev) is how much the team gains per 3-pointer made
# -ev is how much the team loses per 3-pointer missed

#### Top 40 Seasons of All-Time

In [37]:
# sort dataframe by 3PAd
tp=tp.sort_values('3PAd', ascending=False)

# reset index
tp = tp.reset_index(drop=True)

# start index at 1 instead of 0
tp.index = tp.index + 1

# display top 40 seasons
with pd.option_context("display.max_rows", 40): display(tp.head(40))

Year   Tm           Player     G      MP     PTS     3P    3PA    3P%  \
1   2016.0  GSW    Stephen Curry  79.0  2700.0  2375.0  402.0  886.0  0.454   
2   2015.0  ATL      Kyle Korver  75.0  2418.0   911.0  221.0  449.0  0.492   
3   2013.0  GSW    Stephen Curry  78.0  2983.0  1786.0  272.0  600.0  0.453   
4   2015.0  GSW    Stephen Curry  80.0  2613.0  1900.0  286.0  646.0  0.443   
5   2018.0  GSW    Stephen Curry  51.0  1631.0  1346.0  212.0  501.0  0.423   
6   2016.0  LAC      J.J. Redick  75.0  2097.0  1226.0  200.0  421.0  0.475   
7   2017.0  GSW    Stephen Curry  79.0  2638.0  1999.0  324.0  789.0  0.411   
8   2002.0  MIL        Ray Allen  69.0  2525.0  1503.0  229.0  528.0  0.434   
9   2014.0  ATL      Kyle Korver  71.0  2408.0   850.0  185.0  392.0  0.472   
10  1997.0  CHH        Glen Rice  79.0  3362.0  2115.0  207.0  440.0  0.470   
11  2018.0  GSW    Klay Thompson  73.0  2506.0  1461.0  229.0  520.0  0.440   
12  2015.0  GSW    Klay Thompson  77.0  2455.0  1668.0  239.0  545.0  0.439   
13  2016.0  GSW    Klay Thompson  80.0  2666.0  1771.0  276.0  650.0  0.425   
14  2008.0  NOH  Peja Stojakovic  77.0  2711.0  1263.0  231.0  524.0  0.441   
15  2012.0  NYK      Steve Novak  54.0  1020.0   477.0  133.0  282.0  0.472   
16  2014.0  GSW    Stephen Curry  78.0  2846.0  1873.0  261.0  615.0  0.424   
17  1996.0  ORL     Dennis Scott  82.0  3041.0  1431.0  267.0  628.0  0.425   
18  2013.0  ATL      Kyle Korver  74.0  2259.0   810.0  189.0  414.0  0.457   
19  2004.0  SAC  Peja Stojakovic  81.0  3264.0  1964.0  240.0  554.0  0.433   
20  1995.0  PHI      Dana Barros  82.0  3318.0  1686.0  197.0  425.0  0.464   
21  2017.0  GSW    Klay Thompson  78.0  2649.0  1742.0  268.0  647.0  0.414   
22  1996.0  SAC  Mitch Richmond*  81.0  2946.0  1872.0  225.0  515.0  0.437   
23  2006.0  SEA        Ray Allen  78.0  3022.0  1955.0  269.0  653.0  0.412   
24  2005.0  PHO      Joe Johnson  82.0  3240.0  1400.0  177.0  370.0  0.478   
25  2008.0  PHO       Steve Nash  81.0  2780.0  1371.0  179.0  381.0  0.470   
26  2017.0  TOT      Kyle Korver  67.0  1753.0   678.0  162.0  359.0  0.451   
27  2005.0  MIA      Damon Jones  82.0  2576.0   955.0  225.0  521.0  0.432   
28  1997.0  IND   Reggie Miller*  81.0  2966.0  1751.0  229.0  536.0  0.427   
29  2017.0  TOR       Kyle Lowry  60.0  2244.0  1344.0  193.0  468.0  0.412   
30  2006.0  PHO        Raja Bell  79.0  2959.0  1162.0  197.0  446.0  0.442   
31  2014.0  DAL    Jose Calderon  81.0  2468.0   925.0  191.0  425.0  0.449   
32  2012.0  BOS        Ray Allen  46.0  1565.0   655.0  106.0  234.0  0.453   
33  2015.0  LAC      J.J. Redick  78.0  2409.0  1277.0  200.0  458.0  0.437   
34  1996.0  WSB       Tim Legler  77.0  1775.0   726.0  128.0  245.0  0.522   
35  2018.0  UTA       Joe Ingles  82.0  2578.0   940.0  204.0  464.0  0.440   
36  1989.0  SEA       Dale Ellis  82.0  3190.0  2253.0  162.0  339.0  0.478   
37  2015.0  NOP      Eric Gordon  61.0  2018.0   818.0  141.0  315.0  0.448   
38  2017.0  LAC      J.J. Redick  78.0  2198.0  1173.0  201.0  468.0  0.429   
39  2009.0  IND      Troy Murphy  73.0  2482.0  1041.0  161.0  358.0  0.450   
40  2007.0  MIA     Jason Kapono  67.0  1767.0   730.0  108.0  210.0  0.514   

        AM3A      EM3A       AM3P      EM3P       3PG       3PAG     3PMiG  \
1   3.047404  1.523702   6.716418  3.358209  5.088608  11.215190  6.126582   
2   5.385301  2.692650  10.941176  5.470588  2.946667   5.986667  3.040000   
3   4.971667  2.485833  10.966912  5.483456  3.487179   7.692308  4.205128   
4   4.044892  2.022446   9.136364  4.568182  3.575000   8.075000  4.500000   
5   3.255489  1.627745   7.693396  3.846698  4.156863   9.823529  5.666667   
6   4.980998  2.490499  10.485000  5.242500  2.666667   5.613333  2.946667   
7   3.343473  1.671736   8.141975  4.070988  4.101266   9.987342  5.886076   
8   4.782197  2.391098  11.026201  5.513100  3.318841   7.652174  4.333333   
9   6.142857  3.071429  13.016216  6.508108  2.605634   5.521127

3PAd Statistical Notes:<ul>
    <li> Steph Curry's legendary MVP season is heads and shoulders above the rest, and he dominates the list.</li> 
    <li> The metric does an adequate job of comparing 3-point shooters over the years. </li>
     <li> Different expected values will produce different lists. </li>
    <li> The metric has real meaning. It conveys the points a team gains per game by the player shooting 3-pointers. </li>
    </ul>

#### 2018 League Leaders

In [16]:
# Create 2018 dataframe
tp_2018 = tp[tp['Year']==2018.0]

# Show top 20 3PAd
tp_2018.head(20)

Year   Tm            Player     G      MP     PTS     3P    3PA  \
24842  2018.0  GSW     Stephen Curry  51.0  1631.0  1346.0  212.0  501.0   
25301  2018.0  GSW     Klay Thompson  73.0  2506.0  1461.0  229.0  520.0   
24991  2018.0  UTA        Joe Ingles  82.0  2578.0   940.0  204.0  464.0   
25231  2018.0  PHI       J.J. Redick  70.0  2116.0  1198.0  193.0  460.0   
25051  2018.0  CLE       Kyle Korver  73.0  1574.0   672.0  164.0  376.0   
24908  2018.0  OKC       Paul George  79.0  2891.0  1734.0  244.0  608.0   
24869  2018.0  GSW      Kevin Durant  68.0  2325.0  1792.0  173.0  413.0   
24994  2018.0  BOS      Kyrie Irving  60.0  1931.0  1466.0  166.0  407.0   
24784  2018.0  DET    Reggie Bullock  62.0  1732.0   698.0  125.0  281.0   
25082  2018.0  TOR        Kyle Lowry  78.0  2510.0  1267.0  238.0  596.0   
24963  2018.0  SAC       Buddy Hield  80.0  2024.0  1079.0  176.0  408.0   
25305  2018.0  DET  Anthony Tolliver  79.0  1757.0   703.0  159.0  365.0   
25081  2018.0  CLE        Kevin Love  59.0  1651.0  1039.0  137.0  330.0   
24874  2018.0  MIA   Wayne Ellington  77.0  2041.0   864.0  227.0  579.0   
24941  2018.0  DET     Tobias Harris  48.0  1563.0   868.0  113.0  276.0   
25216  2018.0  WAS       Otto Porter  77.0  2432.0  1134.0  138.0  313.0   
24940  2018.0  TOT     Tobias Harris  80.0  2668.0  1486.0  183.0  445.0   
24822  2018.0  IND   Darren Collison  69.0  2018.0   855.0   96.0  205.0   
24939  2018.0  BRK        Joe Harris  78.0  1975.0   846.0  150.0  358.0   
24844  2018.0  PHO      Troy Daniels  79.0  1622.0   703.0  183.0  458.0   

         3P%      AM3A      EM3A       AM3P       EM3P       3PG      3PAG  \
24842  0.423  3.255489  1.627745   7.693396   3.846698  4.156863  9.823529   
25301  0.440  4.819231  2.409615  10.943231   5.471616  3.136986  7.123288   
24991  0.440  5.556034  2.778017  12.637255   6.318627  2.487805  5.658537   
25231  0.420  4.600000  2.300000  10.963731   5.481865  2.757143  6.571429   
25051  0.436  4.186170  2.093085   9.597561   4.798780  2.246575  5.150685   
24908  0.401  4.754934  2.377467  11.848361   5.924180  3.088608  7.696203   
24869  0.419  5.629540  2.814770  13.439306   6.719653  2.544118  6.073529   
24994  0.408  4.744472  2.372236  11.632530   5.816265  2.766667  6.783333   
24784  0.445  6.163701  3.081851  13.856000   6.928000  2.016129  4.532258   
25082  0.399  4.211409  2.105705  10.546218   5.273109  3.051282  7.641026   
24963  0.431  4.960784  2.480392  11.500000   5.750000  2.200000  5.100000   
25305  0.436  4.813699  2.406849  11.050314   5.525157  2.012658  4.620253   
25081  0.415  5.003030  2.501515  12.051095   6.025547  2.322034  5.593220   
24874  0.392  3.525043  1.762522   8.991189   4.495595  2.948052  7.519481   
24941  0.409  5.663043  2.831522  13.831858   6.915929  2.354167  5.750000   
25216  0.441  7.769968  3.884984  17.623188   8.811594  1.792208  4.064935   
24940  0.411  5.995506  2.997753  14.579235   7.289617  2.287500  5.562500   
24822  0.468  9.843902  4.921951  21.020833  10.510417  1.391304  2.971014   
24939  0.419  5.516760  2.758380  13.166667   6.583333  1.923077  4.589744   
24844  0.400  3.541485  1.770742   8.863388   4.431694  2.316456  5.797468   

          3PMiG      3PAd  
24842  5.666667  2.647059  
25301  3.986301  2.287671  
24991  3.170732  1.804878  
25231  3.814286  1.700000  
25051  2.904110  1.589041  
24908  4.607595  1.569620  
24869  3.529412  1.558824  
24994  4.016667  1.516667  
24784  2.516129  1.516129  
25082  4.589744  1.512821  
24963  2.900000  1.500000  
25305  2.607595  1.417722  
25081  3.271186  1.372881  
24874  4.571429  1.324675  
24941  3.395833  1.312500  
25216  2.272727  1.311688  
24940  3.275000  1.300000  
24822  1.579710  1.202899  
24939  2.666667  1.179487  
24844  3.481013  1.151899

The Golden State Warriors dominate the list. What about the Houston Rockets? They have the reputation of being a great 3-point shooting team. Let's make a comparison.

#### Warriors v Rockets

In [17]:
# Create 2018 dataframe for GSW and HOU
tp_2018_GSW_HOU = tp_2018[(tp_2018['Tm']=='GSW') | (tp_2018['Tm']=='HOU')]

# Display dataframe
tp_2018_GSW_HOU

Year   Tm            Player     G      MP     PTS     3P    3PA  \
24842  2018.0  GSW     Stephen Curry  51.0  1631.0  1346.0  212.0  501.0   
25301  2018.0  GSW     Klay Thompson  73.0  2506.0  1461.0  229.0  520.0   
24869  2018.0  GSW      Kevin Durant  68.0  2325.0  1792.0  173.0  413.0   
24932  2018.0  HOU      James Harden  72.0  2551.0  2191.0  265.0  722.0   
25198  2018.0  HOU        Chris Paul  58.0  1847.0  1081.0  144.0  379.0   
24704  2018.0  HOU     Ryan Anderson  66.0  1725.0   617.0  131.0  339.0   
24710  2018.0  HOU      Trevor Ariza  67.0  2269.0   782.0  170.0  462.0   
24915  2018.0  HOU       Eric Gordon  69.0  2154.0  1243.0  218.0  608.0   
25372  2018.0  GSW        Nick Young  80.0  1393.0   581.0  123.0  326.0   
25307  2018.0  HOU       P.J. Tucker  82.0  2281.0   502.0  115.0  310.0   
25098  2018.0  HOU  Luc Mbah a Moute  61.0  1564.0   460.0   63.0  173.0   
24987  2018.0  GSW    Andre Iguodala  64.0  1622.0   384.0   33.0  117.0   
24923  2018.0  GSW    Draymond Green  70.0  2287.0   773.0   77.0  256.0   
25101  2018.0  GSW     Patrick McCaw  57.0   961.0   229.0   19.0   80.0   

         3P%       AM3A      EM3A       AM3P       EM3P       3PG       3PAG  \
24842  0.423   3.255489  1.627745   7.693396   3.846698  4.156863   9.823529   
25301  0.440   4.819231  2.409615  10.943231   5.471616  3.136986   7.123288   
24869  0.419   5.629540  2.814770  13.439306   6.719653  2.544118   6.073529   
24932  0.367   3.533241  1.766620   9.626415   4.813208  3.680556  10.027778   
25198  0.380   4.873351  2.436675  12.826389   6.413194  2.482759   6.534483   
24704  0.386   5.088496  2.544248  13.167939   6.583969  1.984848   5.136364   
24710  0.368   4.911255  2.455628  13.347059   6.673529  2.537313   6.895522   
24915  0.359   3.542763  1.771382   9.880734   4.940367  3.159420   8.811594   
25372  0.377   4.273006  2.136503  11.325203   5.662602  1.537500   4.075000   
25307  0.371   7.358065  3.679032  19.834783   9.917391  1.402439   3.780488   
25098  0.364   9.040462  4.520231  24.825397  12.412698  1.032787   2.836066   
24987  0.282  13.863248  6.931624  49.151515  24.575758  0.515625   1.828125   
24923  0.301   8.933594  4.466797  29.701299  14.850649  1.100000   3.657143   
25101  0.238  12.012500  6.006250  50.578947  25.289474  0.333333   1.403509   

          3PMiG      3PAd  
24842  5.666667  2.647059  
25301  3.986301  2.287671  
24869  3.529412  1.558824  
24932  6.347222  1.013889  
25198  4.051724  0.913793  
24704  3.151515  0.818182  
24710  4.358209  0.716418  
24915  5.652174  0.666667  
25372  2.537500  0.537500  
25307  2.378049  0.426829  
25098  1.803279  0.262295  
24987  1.312500 -0.281250  
24923  2.557143 -0.357143  
25101  1.070175 -0.403509

Even though Golden State has the best shooters, it's not clear which team is the best at shooting 3's. This is easily resolved by summing 3PAd.

In [24]:
tp_2018_GSW_HOU['3PAd/T'] = tp_2018_GSW_HOU.groupby('Tm')['3PAd'].transform('sum')

/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [25]:
# Display updated dataframe
tp_2018_GSW_HOU

Year   Tm            Player     G      MP     PTS     3P    3PA  \
24842  2018.0  GSW     Stephen Curry  51.0  1631.0  1346.0  212.0  501.0   
25301  2018.0  GSW     Klay Thompson  73.0  2506.0  1461.0  229.0  520.0   
24869  2018.0  GSW      Kevin Durant  68.0  2325.0  1792.0  173.0  413.0   
24932  2018.0  HOU      James Harden  72.0  2551.0  2191.0  265.0  722.0   
25198  2018.0  HOU        Chris Paul  58.0  1847.0  1081.0  144.0  379.0   
24704  2018.0  HOU     Ryan Anderson  66.0  1725.0   617.0  131.0  339.0   
24710  2018.0  HOU      Trevor Ariza  67.0  2269.0   782.0  170.0  462.0   
24915  2018.0  HOU       Eric Gordon  69.0  2154.0  1243.0  218.0  608.0   
25372  2018.0  GSW        Nick Young  80.0  1393.0   581.0  123.0  326.0   
25307  2018.0  HOU       P.J. Tucker  82.0  2281.0   502.0  115.0  310.0   
25098  2018.0  HOU  Luc Mbah a Moute  61.0  1564.0   460.0   63.0  173.0   
24987  2018.0  GSW    Andre Iguodala  64.0  1622.0   384.0   33.0  117.0   
24923  2018.0  GSW    Draymond Green  70.0  2287.0   773.0   77.0  256.0   
25101  2018.0  GSW     Patrick McCaw  57.0   961.0   229.0   19.0   80.0   

         3P%       AM3A      EM3A       AM3P       EM3P       3PG       3PAG  \
24842  0.423   3.255489  1.627745   7.693396   3.846698  4.156863   9.823529   
25301  0.440   4.819231  2.409615  10.943231   5.471616  3.136986   7.123288   
24869  0.419   5.629540  2.814770  13.439306   6.719653  2.544118   6.073529   
24932  0.367   3.533241  1.766620   9.626415   4.813208  3.680556  10.027778   
25198  0.380   4.873351  2.436675  12.826389   6.413194  2.482759   6.534483   
24704  0.386   5.088496  2.544248  13.167939   6.583969  1.984848   5.136364   
24710  0.368   4.911255  2.455628  13.347059   6.673529  2.537313   6.895522   
24915  0.359   3.542763  1.771382   9.880734   4.940367  3.159420   8.811594   
25372  0.377   4.273006  2.136503  11.325203   5.662602  1.537500   4.075000   
25307  0.371   7.358065  3.679032  19.834783   9.917391  1.402439   3.780488   
25098  0.364   9.040462  4.520231  24.825397  12.412698  1.032787   2.836066   
24987  0.282  13.863248  6.931624  49.151515  24.575758  0.515625   1.828125   
24923  0.301   8.933594  4.466797  29.701299  14.850649  1.100000   3.657143   
25101  0.238  12.012500  6.006250  50.578947  25.289474  0.333333   1.403509   

          3PMiG      3PAd    3PAd/T  
24842  5.666667  2.647059  5.989152  
25301  3.986301  2.287671  5.989152  
24869  3.529412  1.558824  5.989152  
24932  6.347222  1.013889  4.818073  
25198  4.051724  0.913793  4.818073  
24704  3.151515  0.818182  4.818073  
24710  4.358209  0.716418  4.818073  
24915  5.652174  0.666667  4.818073  
25372  2.537500  0.537500  5.989152  
25307  2.378049  0.426829  4.818073  
25098  1.803279  0.262295  4.818073  
24987  1.312500 -0.281250  5.989152  
24923  2.557143 -0.357143  5.989152  
25101  1.070175 -0.403509  5.989152

Golden State is the clear winner. They gain 5.99 points per game by shooting 3s while Houston gains 4.81.

## Fun with 3PAd

Who had the best 3-point seasons before the year 2000? Let's find out.

#### Best of the 90s

In [46]:
tp_before2000 = tp[(tp['Year']<2000) & (tp['Year']>1989)]

with pd.option_context("display.max_rows", 25): display(tp_before2000.head(25))

Year   Tm           Player     G      MP     PTS     3P    3PA    3P%  \
10   1997.0  CHH        Glen Rice  79.0  3362.0  2115.0  207.0  440.0  0.470   
17   1996.0  ORL     Dennis Scott  82.0  3041.0  1431.0  267.0  628.0  0.425   
20   1995.0  PHI      Dana Barros  82.0  3318.0  1686.0  197.0  425.0  0.464   
22   1996.0  SAC  Mitch Richmond*  81.0  2946.0  1872.0  225.0  515.0  0.437   
28   1997.0  IND   Reggie Miller*  81.0  2966.0  1751.0  229.0  536.0  0.427   
34   1996.0  WSB       Tim Legler  77.0  1775.0   726.0  128.0  245.0  0.522   
43   1997.0  SAC  Mitch Richmond*  81.0  3125.0  2095.0  204.0  477.0  0.428   
55   1995.0  ORL     Dennis Scott  62.0  1499.0   802.0  150.0  352.0  0.426   
57   1996.0  CHI       Steve Kerr  82.0  1919.0   688.0  122.0  237.0  0.515   
58   1998.0  CLE    Wesley Person  82.0  3198.0  1204.0  192.0  447.0  0.430   
65   1996.0  NYK     Hubert Davis  74.0  1773.0   789.0  127.0  267.0  0.476   
66   1996.0  DET    Allan Houston  82.0  3072.0  1617.0  191.0  447.0  0.427   
86   1999.0  MIL       Dell Curry  42.0   864.0   423.0   69.0  145.0  0.476   
90   1997.0  MIA    Voshon Lenard  73.0  2111.0   897.0  183.0  442.0  0.414   
91   1995.0  CHH       Dell Curry  69.0  1718.0   935.0  154.0  361.0  0.427   
95   1997.0  DET      Joe Dumars*  79.0  2923.0  1158.0  166.0  384.0  0.432   
99   1996.0  WSB      Brent Price  81.0  2042.0   810.0  139.0  301.0  0.462   
102  1995.0  IND   Reggie Miller*  81.0  2665.0  1588.0  195.0  470.0  0.415   
107  1995.0  ORL    Nick Anderson  76.0  2588.0  1200.0  179.0  431.0  0.415   
108  1996.0  CHH        Glen Rice  79.0  3142.0  1710.0  171.0  403.0  0.424   
109  1997.0  DET      Terry Mills  79.0  1997.0   857.0  175.0  415.0  0.422   
121  1998.0  IND   Reggie Miller*  81.0  2795.0  1578.0  164.0  382.0  0.429   
122  1998.0  SEA       Dale Ellis  79.0  1939.0   934.0  127.0  274.0  0.464   
123  1999.0  SEA       Dale Ellis  48.0  1232.0   495.0   94.0  217.0  0.433   
130  1996.0  SAS     Chuck Person  80.0  2131.0   873.0  190.0  463.0  0.410   

         AM3A      EM3A       AM3P      EM3P       3PG      3PAG     3PMiG  \
10   7.640909  3.820455  16.241546  8.120773  2.620253  5.569620  2.949367   
17   4.842357  2.421178  11.389513  5.694757  3.256098  7.658537  4.402439   
20   7.807059  3.903529  16.842640  8.421320  2.402439  5.182927  2.780488   
22   5.720388  2.860194  13.093333  6.546667  2.777778  6.358025  3.580247   
28   5.533582  2.766791  12.951965  6.475983  2.827160  6.617284  3.790123   
34   7.244898  3.622449  13.867188  6.933594  1.662338  3.181818  1.519481   
43   6.551363  3.275681  15.318627  7.659314  2.518519  5.888889  3.370370   
55   4.258523  2.129261   9.993333  4.996667  2.419355  5.677419  3.258065   
57   8.097046  4.048523  15.729508  7.864754  1.487805  2.890244  1.402439   
58   7.154362  3.577181  16.656250  8.328125  2.341463  5.451220  3.109756   
65   6.640449  3.320225  13.960630  6.980315  1.716216  3.608108  1.891892   
66   6.872483  3.436242  16.083770  8.041885  2.329268  5.451220  3.121951   
86   5.958621  2.979310  12.521739  6.260870  1.642857  3.452381  1.809524   
90   4.776018  2.388009  11.535519  5.767760  2.506849  6.054795  3.547945   
91   4.759003  2.379501  11.155844  5.577922  2.231884  5.231884  3.000000   
95   7.611979  3.805990  17.608434  8.804217  2.101266  4.860759  2.759494   
99   6.784053  3.392027  14.690647  7.345324  1.716049  3.716049  2.000000   
102  5.670213  2.835106  13.666667  6.833333  2.407407  5.802469  3.395062   
107  6.004640  3.002320  14.458101  7.229050  2.355263  5.671053  3.315789   
108  7.796526  3.898263  18.374269  9.187135  2.164557  5.101266  2.936709   
109  4.812048  2.406024  11.411429  5.705714  2.215190  5.253165  3.037975   
121  7.316754  3.658377  17.042683  8.521341  2.024691  4.716049  2.691358   
122  7.076642  3.538321  15.267717  7.633858  1.607595  3.468354  1.860759   
123  5.677419  2.838710  13.106383  6.553191  1.958333

The big 90s shooters. Glen Rice, Reggie Miller, Dell Curry, Dennis Scott, Mitch Ritchmond, Dale Ellis. What about the 80s?

#### Best of the 80s

In [45]:
tp_before90 = tp[tp['Year']<1990]

with pd.option_context("display.max_rows", 25): display(tp_before90.head(25))

Year   Tm            Player     G      MP     PTS     3P    3PA  \
36    1989.0  SEA        Dale Ellis  82.0  3190.0  2253.0  162.0  339.0   
163   1988.0  TOT      Craig Hodges  66.0  1445.0   629.0   86.0  175.0   
262   1988.0  MIL      Craig Hodges  43.0   983.0   397.0   55.0  118.0   
272   1988.0  BOS       Danny Ainge  81.0  3018.0  1270.0  148.0  357.0   
427   1989.0  CHI      Craig Hodges  49.0  1112.0   490.0   71.0  168.0   
445   1989.0  CLE        Mark Price  75.0  2728.0  1414.0   93.0  211.0   
465   1987.0  BOS       Danny Ainge  71.0  2499.0  1053.0   85.0  192.0   
505   1986.0  MIL      Craig Hodges  66.0  1739.0   716.0   73.0  162.0   
533   1988.0  CLE        Mark Price  80.0  2626.0  1279.0   72.0  148.0   
566   1988.0  SEA        Dale Ellis  75.0  2790.0  1938.0  107.0  259.0   
648   1989.0  SAC   Harold Pressley  80.0  2257.0   981.0  119.0  295.0   
664   1989.0  MIA      Jon Sundvold  68.0  1338.0   709.0   48.0   92.0   
667   1989.0  TOT      Craig Hodges  59.0  1204.0   529.0   75.0  180.0   
680   1988.0  BOS       Larry Bird*  76.0  2965.0  2275.0   98.0  237.0   
742   1989.0  NYK      Trent Tucker  81.0  1824.0   687.0  118.0  296.0   
823   1989.0  IND    Reggie Miller*  74.0  2536.0  1181.0   98.0  244.0   
907   1986.0  BOS       Larry Bird*  82.0  3113.0  2115.0   82.0  194.0   
917   1980.0  BOS        Chris Ford  73.0  2115.0   816.0   70.0  164.0   
953   1987.0  NYK      Trent Tucker  70.0  1691.0   795.0   68.0  161.0   
970   1988.0  PHI  Gerald Henderson  69.0  1436.0   581.0   67.0  159.0   
972   1987.0  BOS       Larry Bird*  74.0  3005.0  2076.0   90.0  225.0   
1009  1989.0  PHI    Hersey Hawkins  79.0  2577.0  1196.0   71.0  166.0   
1020  1989.0  TOT       Danny Ainge  73.0  2377.0  1281.0  116.0  305.0   
1025  1986.0  CLE          World B.  75.0  2535.0  1754.0   71.0  169.0   
1026  1988.0  TOT  Gerald Henderson  75.0  1505.0   595.0   69.0  163.0   

        3P%       AM3A      EM3A       AM3P       EM3P       3PG      3PAG  \
36    0.478   9.410029  4.705015  19.691358   9.845679  1.975610  4.134146   
163   0.491   8.257143  4.128571  16.802326   8.401163  1.303030  2.651515   
262   0.466   8.330508  4.165254  17.872727   8.936364  1.279070  2.744186   
272   0.415   8.453782  4.226891  20.391892  10.195946  1.827160  4.407407   
427   0.423   6.619048  3.309524  15.661972   7.830986  1.448980  3.428571   
445   0.441  12.928910  6.464455  29.333333  14.666667  1.240000  2.813333   
465   0.443  13.015625  6.507812  29.400000  14.700000  1.197183  2.704225   
505   0.451  10.734568  5.367284  23.821918  11.910959  1.106061  2.454545   
533   0.486  17.743243  8.871622  36.472222  18.236111  0.900000  1.850000   
566   0.413  10.772201  5.386100  26.074766  13.037383  1.426667  3.453333   
648   0.403   7.650847  3.825424  18.966387   9.483193  1.487500  3.687500   
664   0.522  14.543478  7.271739  27.875000  13.937500  0.705882  1.352941   
667   0.417   6.688889  3.344444  16.053333   8.026667  1.271186  3.050847   
680   0.414  12.510549  6.255274  30.255102  15.127551  1.289474  3.118421   
742   0.399   6.162162  3.081081  15.457627   7.728814  1.456790  3.654321   
823   0.402  10.393443  5.196721  25.877551  12.938776  1.324324  3.297297   
907   0.423  16.046392  8.023196  37.963415  18.981707  1.000000  2.365854   
917   0.427  12.896341  6.448171  30.214286  15.107143  0.958904  2.246575   
953   0.422  10.503106  5.251553  24.867647  12.433824  0.971429  2.300000   
970   0.421   9.031447  4.515723  21.432836  10.716418  0.971014  2.304348   
972   0.400  13.355556  6.677778  33.388889  16.694444  1.216216  3.040541   
1009  0.428  15.524096  7.762048  36.295775  18.147887  0.898734  2.101266   
1020  0.380   7.793443  3.896721  20.491379  10.245690  1.589041  4.178082   
1025  0.420  15.000000  7.500000  35.704225  17.852113  0.946667  2.253333   
1026  0.423   9.233129  4.616564  21.811594  10.905797  0.920000  2.173333   

         3PMiG      3PAd  
3

Craig Hodges. Mark Price. Larry Bird. More Dale Ellis. The first 3-point shot was made in 1979, so we can't go back much further.

#### Best Career Totals

In [52]:
tp['3PNG/C'] tp['3P'] * (3 - ev) - (tp['3PA'] - tp['3P']) * ev

tp_player = tp.groupby('Player')['3PNG/C'].sum()
tp_player = tp_player.sort_values(ascending=False)
with pd.option_context("display.max_rows", 25): display(tp_player.head(25))

Player
Ray Allen           21.905276
Kyle Korver         21.289620
Stephen Curry       19.562224
Reggie Miller*      15.245892
Steve Nash          14.655158
Chauncey Billups    12.766855
Klay Thompson       12.703306
J.J. Redick         12.424284
Dale Ellis          12.127907
Mike Miller         11.625611
Peja Stojakovic     11.457409
Jason Terry         10.602291
Vince Carter        10.519366
Rashard Lewis       10.357993
Brent Barry         10.145916
Glen Rice            9.934763
Dirk Nowitzki        9.782523
Dell Curry           9.485008
J.R. Smith           9.370183
Wesley Person        9.033540
Jose Calderon        9.003851
Hubert Davis         8.865226
Allan Houston        8.862571
Paul Pierce          8.665088
Mike Bibby           8.587188
Name: 3PAd, dtype: float64

In [55]:
# Create new category, use totals instead of per game
tp['3PNG/C'] = tp['3P'] * (3 - ev) - (tp['3PA'] - tp['3P']) * ev

# group by player, and sum over their career
tp_player = tp.groupby('Player')['3PNG/C'].sum()

# order from the top
tp_player = tp_player.sort_values(ascending=False)

# display top 25
tp_player.head(25)

Player
Ray Allen           1548.0
Kyle Korver         1540.0
Stephen Curry       1463.0
Reggie Miller*      1194.0
Steve Nash          1100.0
Klay Thompson        980.0
Chauncey Billups     933.0
Dale Ellis           891.0
J.J. Redick          877.0
Peja Stojakovic      835.0
Jason Terry          822.0
Mike Miller          808.0
Glen Rice            780.0
Vince Carter         762.0
Rashard Lewis        758.0
Dirk Nowitzki        749.0
Brent Barry          727.0
Wesley Person        662.0
J.R. Smith           658.0
Allan Houston        655.0
Mike Bibby           639.0
Dell Curry           637.0
Joe Johnson          613.0
Paul Pierce          611.0
Hubert Davis         598.0
Name: 3PNG/C, dtype: float64

What about best teams of all-time?

In [ ]:
tp_player = tp.groupby('Team')['3PAd'].sum()
tp_player = tp_player.sort_values(ascending=False)
with pd.option_context("display.max_rows", 25): display(tp_player.head(25))

## Conclusion

3PAd is a fun, informative and comparative statistic used to rank 3-point shooters. It is a single metric that combines the number of 3-pointers made with a players 3-point percentage. It stands up well to general expectations about who the best 3-point shooters are. It can be used to compare teams, and players from any year and different eras. 